### PR TITLE
Introduce cpp micro-benchmarks

### PR DESCRIPTION
**Pull request by dszwicht**
This PR introduces C++ benchmarking infrastructure ([GoogleBenchmark](https://github.com/google/benchmark)). As `pyg-lib` can be used as a standalone C++ library it makes sense to have micro-benchmarks that would benchmark not only pieces of code with a Python binding. Additionally, an exemplary benchmark was provided for `pyg::sampler::Mapper`.

Output for provided benchmark looks like follows:
```
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                       Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:1024000/num_entries:-1    7394664 ns      7392668 ns           91 Insertion Fail Rate [%]=25.0501
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:1152000/num_entries:-1    8760396 ns      8757657 ns           77 Insertion Fail Rate [%]=25.0199
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:1280000/num_entries:-1    8948630 ns      8945524 ns           76 Insertion Fail Rate [%]=25.0318
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:1408000/num_entries:-1   10103691 ns     10099494 ns           68 Insertion Fail Rate [%]=25.0122
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:1536000/num_entries:-1   11302446 ns     11296285 ns           61 Insertion Fail Rate [%]=24.993
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:1664000/num_entries:-1   12936696 ns     12929287 ns           54 Insertion Fail Rate [%]=25.0384
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:1792000/num_entries:-1   14151064 ns     14144356 ns           49 Insertion Fail Rate [%]=25.0186
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:1920000/num_entries:-1   15588397 ns     15576363 ns           45 Insertion Fail Rate [%]=25.0133
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap/num_nodes:2048000/num_entries:-1   17158246 ns     17144487 ns           41 Insertion Fail Rate [%]=25.007
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap_BigO                                   7.77 N          7.77 N    
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithFlatHashMap_RMS                                       6 %             6 %    
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:1024000/num_entries:1024000    6776261 ns      6773525 ns          103 Insertion Fail Rate [%]=25.0364
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:1152000/num_entries:1152000    7983127 ns      7979648 ns           89 Insertion Fail Rate [%]=24.9776
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:1280000/num_entries:1280000    8971134 ns      8965624 ns           78 Insertion Fail Rate [%]=25.0282
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:1408000/num_entries:1408000   10156930 ns     10151395 ns           68 Insertion Fail Rate [%]=25.0169
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:1536000/num_entries:1536000   11407066 ns     11399957 ns           61 Insertion Fail Rate [%]=25.004
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:1664000/num_entries:1664000   12593205 ns     12587564 ns           55 Insertion Fail Rate [%]=24.9894
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:1792000/num_entries:1792000   14226021 ns     14215225 ns           49 Insertion Fail Rate [%]=25.0002
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:1920000/num_entries:1920000   15570986 ns     15562529 ns           44 Insertion Fail Rate [%]=24.9714
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector/num_nodes:2048000/num_entries:2048000   17869559 ns     17851791 ns           39 Insertion Fail Rate [%]=24.996
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector_BigO                                        7.76 N          7.75 N    
BenchmarkMapperCreationAndInsertion<int64_t, int64_t>/WithVector_RMS                                            8 %             8 %
```

### CODE REVIEW

**Comment by dszwicht** [comment]
@rusty1s waiting for some feedback 😉

File: benchmark/csrc/sampler/mapper.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+#include <type_traits>
+#include <vector>
+
+#include <benchmark/benchmark.h>
+
+#include "pyg_lib/csrc/sampler/cpu/mapper.h"
+
+template <typename node_t, typename scalar_t>
+class BenchmarkMapperCreationAndInsertion : public benchmark::Fixture {
+  static_assert(std::is_integral<node_t>::value &&
```
Comment (rusty1s) [inline benchmark/csrc/sampler/mapper.cpp:12]: Do we need this check? I think we can also benchmark `node_t` without being necessarily an integral.

File: benchmark/csrc/sampler/mapper.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                    std::is_integral<scalar_t>::value,
+                "Integral type required for both node_t and scalar_t");
+
+ protected:
+  void SetUp(const benchmark::State& state) override {
+    static std::random_device rd;
```
Comment (rusty1s) [inline benchmark/csrc/sampler/mapper.cpp:18]: Not really familiar with this. Is it deterministic?

File: benchmark/csrc/sampler/mapper.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    const auto num_nodes = state.range(0);
+    nodes_.resize(num_nodes * 2);
+    std::iota(nodes_.begin(), nodes_.begin() + num_nodes, 0);
+    std::iota(nodes_.begin() + num_nodes, nodes_.end(), 0);
+    std::shuffle(nodes_.begin(), nodes_.end(), gen);
+  }
+
+  void TearDown(const benchmark::State& state) override { nodes_.clear(); }
+
+  void PerformTest(size_t num_nodes, size_t num_entries, size_t samples) {
```
Comment (rusty1s) [inline benchmark/csrc/sampler/mapper.cpp:31]: I think we can safely remove the `num_entries` argument here. The `num_entries` argument dictates how many samples we approximately want to map. In my understanding, this benchmark assumes that we map all nodes.

File: benchmark/csrc/sampler/mapper.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    insertion_fail_counter_ = 0;
+    pyg::sampler::Mapper<node_t, scalar_t> mapper(num_nodes, num_entries);
```
Comment (rusty1s) [inline benchmark/csrc/sampler/mapper.cpp:33]:
```suggestion
    pyg::sampler::Mapper<node_t, scalar_t> mapper(num_nodes);
```

**Review by rusty1s — APPROVED** [review]
This is super cool, thanks for adding it!

File: benchmark/csrc/sampler/mapper.cpp (hunk 3 above)
Comment (dszwicht) [reply benchmark/csrc/sampler/mapper.cpp:31]: This benchmark in particular wants to test heuristics for vector usage. We use vector if `num_nodes < 1'000'000` or if `num_entries` is more than 10% of `num_nodes`. By choosing `num_nodes > 1'000'000` and setting explicitly `num_entries` to -1 or `num_nodes` we can test map and vector (respectively) on the same scenarios.

File: benchmark/csrc/sampler/mapper.cpp (hunk 1 above)
Comment (dszwicht) [reply benchmark/csrc/sampler/mapper.cpp:12]: As in below [comment](https://github.com/pyg-team/pyg-lib/pull/101#discussion_r976057274), this is needed to have a single case which works on both map and vector under the hood.

**Comment by dszwicht** [comment]
@rusty1s I replied to all comments, can I merge this PR? 😉

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -31,11 +32,12 @@ mkdir build
 cd build
 export Torch_DIR=`python -c 'import torch;print(torch.utils.cmake_prefix_path)'`
 export CUDA_ARCH_LIST=75
-cmake .. -GNinja -DBUILD_TEST=ON -DWITH_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCH_LIST}
+cmake .. -GNinja -DBUILD_TEST=ON -DBUILD_BENCHMARK=ON -DWITH_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCH_LIST}
 cmake --build .
 ```
 
-If you want to include the test suite, specify `-DBUILD_TEST=ON` (`OFF` by default).
+If you want to include the test suite or benchmarks, specify `-DBUILD_TEST=ON` or `-DBUILD_BENCHMARK=ON` respectively
```
Comment (rusty1s) [inline CONTRIBUTING.md:39]:
```suggestion
If you want to include the test suite or benchmarks, specify `-DBUILD_TEST=ON` or `-DBUILD_BENCHMARK=ON`, respectively
```

File: benchmark/csrc/sampler/mapper.cpp (hunk 2 above)
Comment (rusty1s) [reply benchmark/csrc/sampler/mapper.cpp:18]: Would it make sense to make this deterministic such that different runs are comparable? Not a blocker though.

File: benchmark/csrc/sampler/mapper.cpp (hunk 3 above)
Comment (rusty1s) [reply benchmark/csrc/sampler/mapper.cpp:31]: I see. That makes sense.

**Review by rusty1s — APPROVED** [review]

**Comment by rusty1s** [comment]
@dszwicht Yes, ofc. Please go ahead and merge it if it is ready.

File: benchmark/csrc/sampler/mapper.cpp (hunk 2 above)
Comment (dszwicht) [reply benchmark/csrc/sampler/mapper.cpp]: Yes, just did it.

**Pull request merged**